### PR TITLE
esp32 lock-app mdns server start at connectivity change

### LINKED ISSUE #7705: chip-device-ctrl - resolve command hangs with lock-app
#### Problem
chip-device-ctrl - resolve command hangs with lock-app on esp32
 * expected behavior : resolve dns and update address of the node in the device controller after command
      `resolve 0 135246`
 * actual behavior - resolve command hangs
 * system configuration
    - platform : macOS
    - kit : esp32
 *  logs
    - device side : [device-logs.txt](https://github.com/project-chip/connectedhomeip/files/6667681/device-logs.txt)

    - python controller side : [chip-device-ctrl.txt](https://github.com/project-chip/connectedhomeip/files/6667678/chip-device-ctrl.txt)

### PR DESCRIPTION
**Pull request by DamKast**
#### Problem
What is being fixed?  Examples:
* Fixes #7705 chip-device-ctrl - resolve command hangs with lock-app.

#### Change overview
esp32 lock-app mdns server start at connectivity change

#### Testing
How was this tested? (at least one bullet point required)
* using lock-app on esp32

### CODE REVIEW

**Comment by dhrishi** [comment]
Hi @DamKast Thanks for the fix! The same issue is also seen with `examples/temperature-measurement-app/esp32` as well. Would you mind making the exact same changes there?

File: examples/lock-app/esp32/main/DeviceCallbacks.cpp [hunk 1]
```diff
@@ -51,6 +52,18 @@ void DeviceCallbacks::DeviceEventCallback(const ChipDeviceEvent * event, intptr_
     case DeviceEventType::kSessionEstablished:
         OnSessionEstablished(event);
         break;
+
+    case DeviceEventType::kInterfaceIpAddressChanged:
```
Comment (msandstedt) [inline examples/lock-app/esp32/main/DeviceCallbacks.cpp:56]: Would it be simpler or clearer to place all `chip::app::Mdns::StartServer();` calls in `DeviceEventCallback`?

**Review by gjc13 — APPROVED** [review]

File: examples/lock-app/esp32/main/DeviceCallbacks.cpp (hunk 1 above)
Comment (DamKast) [reply examples/lock-app/esp32/main/DeviceCallbacks.cpp:56]: This is how it is done for the all-clusters-app example. So I am not sure that moving it in `DeviceEventCallback` would be clearer.

**Comment by klem-everywhere** [comment]
Me and @DamKast are looking into if it makes more sense to move it into `src/app/server/Server.cpp` or even `src/app/server/Mdns.cpp` to be more generic.
Probably an `#elsif` in 
```
// ESP32 examples have a custom logic for enabling DNS-SD
#if CHIP_DEVICE_CONFIG_ENABLE_MDNS && !CHIP_DEVICE_LAYER_TARGET_ESP32
    app::Mdns::StartServer();
#endif
```

Or in `app::Mdns::StartServer()`.

Thoughts?

**Review by andy31415 — APPROVED** [review]

**Comment by klem-everywhere** [comment]
To elaborate my last comment, what do you guys think of this approach instead?
I can get @DamKast to polish it for this PR if it's good.

```diff
diff --git a/src/app/server/Mdns.cpp b/src/app/server/Mdns.cpp
index 3bab44a9..157c4b5d 100644
--- a/src/app/server/Mdns.cpp
+++ b/src/app/server/Mdns.cpp
@@ -272,6 +272,21 @@ void StartServer()
     }
 }
 
+void NetworkChangeEventHandler(const chip::DeviceLayer::ChipDeviceEvent * event, intptr_t)
+{
+    VerifyOrReturn(event->Type == chip::DeviceLayer::DeviceEventType::kInternetConnectivityChange);
+
+    if (event->InternetConnectivityChange.IPv4 == chip::DeviceLayer::kConnectivity_Established || event->InternetConnectivityChange.IPv6 == chip::DeviceLayer::kConnectivity_Established)
+    {
+        chip::app::Mdns::StartServer();
+    }
+}
+
+void StartServerOnNetworkChange()
+{
+    chip::DeviceLayer::PlatformMgr().AddEventHandler(NetworkChangeEventHandler, {});
+}
+
 #if CHIP_ENABLE_ROTATING_DEVICE_ID
 CHIP_ERROR GenerateRotatingDeviceId(char rotatingDeviceIdHexBuffer[], size_t rotatingDeviceIdHexBufferSize)
 {
diff --git a/src/app/server/Mdns.h b/src/app/server/Mdns.h
index 7c13968f..4f21dc12 100644
--- a/src/app/server/Mdns.h
+++ b/src/app/server/Mdns.h
@@ -40,6 +40,9 @@ CHIP_ERROR Advertise(bool commissionableNode);
 /// (Re-)starts the minmdns server
 void StartServer();
 
+/// (Re-)starts the minmdns server on Network change
+void StartServerOnIpChange();
+
 CHIP_ERROR GenerateRotatingDeviceId(char rotatingDeviceIdHexBuffer[], size_t rotatingDeviceIdHexBufferSize);
 
 } // namespace Mdns
diff --git a/src/app/server/Server.cpp b/src/app/server/Server.cpp
index 4198882d..466cb508 100644
--- a/src/app/server/Server.cpp
+++ b/src/app/server/Server.cpp
@@ -527,9 +527,10 @@ void InitServer(AppDelegate * delegate)
 #endif
     }
 
-// ESP32 examples have a custom logic for enabling DNS-SD
 #if CHIP_DEVICE_CONFIG_ENABLE_MDNS && !CHIP_DEVICE_LAYER_TARGET_ESP32
     app::Mdns::StartServer();
+#elif CHIP_DEVICE_LAYER_TARGET_ESP32
+    app::Mdns::StartServerOnIpChange();
 #endif
 
     gCallbacks.SetSessionMgr(&gSessions);
```

File: examples/lock-app/esp32/main/DeviceCallbacks.cpp (hunk 1 above)
Comment (msandstedt) [reply examples/lock-app/esp32/main/DeviceCallbacks.cpp:56]: Sure, OK.  I mean it's just moving the calls one function deeper.

No problem.

**Review by msandstedt — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

**Pull request merged**